modules/3d/src/mst.cpp

```diff
@@ -0,0 +1,160 @@
+{
+    if (a.weight != b.weight)
+        return a.weight < b.weight;
+    if (a.source != b.source)
+        return a.source < b.source;
+    return a.target < b.target;
+}
+
+bool buildMSTKruskal(int numNodes,
+                    const std::vector<cv::MSTEdge>& edges,
+                    std::vector<cv::MSTEdge>& resultingEdges)
+{
+    std::vector<cv::MSTEdge> sortedEdges = edges;
+    std::sort(sortedEdges.begin(), sortedEdges.end(), weightComparator);
+    DSU dsu(numNodes);
+
+    for (const auto &e : sortedEdges)
+    {
+        int u = e.source, v = e.target;
+        if (u >= numNodes || v >= numNodes)
+            return false;
+        if (dsu.find(u) != dsu.find(v))
+        {
+            resultingEdges.push_back(e);
+            dsu.unite(u, v);
+        }
+    }
+
+    return true;
+}
+
+bool buildMSTPrim(int numNodes,
+                  const std::vector<cv::MSTEdge>& edges,
+                  std::vector<cv::MSTEdge>& resultingEdges,
+                  int root)
+{
+    std::vector<bool> inMST(numNodes, false);
+    std::vector<std::vector<cv::MSTEdge>> adj(numNodes);
+    for (const auto& e : edges)
+    {
+        int u = e.source, v = e.target;
+        if (u >= numNodes || v >= numNodes)
+            return false;
+        adj[u].push_back({u, v, e.weight});
+        adj[v].push_back({v, u, e.weight});
+    }
+
+    using HeapElem = std::tuple<double, int, int>; // (weight, from, to)
+    std::priority_queue<HeapElem, std::vector<HeapElem>, std::greater<HeapElem>> pq;
+
+    inMST[root] = true;
+    for (const auto& e : adj[root])
+    {
+        pq.emplace(e.weight, root, e.target);
+    }
+
+    while (!pq.empty())
+    {
+        auto [w, u, v] = pq.top();
+        pq.pop();
+
+        if (inMST[v])
+            continue;
+
+        inMST[v] = true;
+        resultingEdges.push_back({u, v, w});
+        for (const auto& e : adj[v])
+        {
+            if (!inMST[e.target])
+                pq.emplace(e.weight, v, e.target);
+        }
+    }
+
+    return true;
+}
+
+} // unamed namespace
+
+namespace cv
+{
+
+bool buildMST(int numNodes,
+              const std::vector<cv::MSTEdge>& inputEdges,
+              std::vector<cv::MSTEdge>& resultingEdges,
+              MSTAlgorithm algorithm,
+              int root)
+{
+    CV_TRACE_FUNCTION();
+
+    resultingEdges.clear();
+    if (numNodes <= 0 || inputEdges.empty() || root >= numNodes)
+        return false;
+
+    bool result = false;
+    switch (algorithm)
+    {
+        case MST_PRIM:
+            result = buildMSTPrim(numNodes, inputEdges, resultingEdges, root);
+            break;
+        case MST_KRUSKAL:
+            result = buildMSTKruskal(numNodes, inputEdges, resultingEdges);
+            break;
+        default:
+            CV_LOG_INFO(NULL, "Invalid MST algorithm specified");
+            result = false;
+    }
+
+    return (result && resultingEdges.size() >= static_cast<size_t>(numNodes - 1));
```

Comment:
It's the only place where false is returned, besides the obvious one. What happens if the MST cannot be built, e.g. the graph consists of 2 split sub-graphs? I would say, it should return false too.